src/jit/hwintrinsicxarch.cpp

```diff
@@ -732,22 +732,33 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
     {
         return impUnsupportedHWIntrinsic(CORINFO_HELP_THROW_PLATFORM_NOT_SUPPORTED, method, sig, mustExpand);
     }
+
+    GenTree* lastOp = nullptr;
+    bool isConstImm = false;
+
     // Avoid checking stacktop for 0-op intrinsics
     if (sig->numArgs > 0 && HWIntrinsicInfo::isImmOp(intrinsic, impStackTop().val))
     {
-        GenTree* lastOp = impStackTop().val;
+        lastOp     = impStackTop().val;
+        isConstImm = lastOp->IsCnsIntOrI();
+
+        if (!isConstImm)
+        {
+            lastOp     = fgMorphTree(lastOp);
```

Comment:
I'm a bit nervous about calling `fgMorphTree` here, as we don't do that anywhere else, and it's unclear whether the importer can handle all of the transformations it might do (e.g. we can't split trees in the importer).
@dotnet/jit-contrib - any thoughts?
My paranoid side says that we should constrain this to something simpler, such as just folding of constants, but maybe that's also introducing needless complexity.

Reply:
Tried to analyze impact of such call on jit state and it seems that besides tree morphing there are no other persistent changes. I fully understand hesitation to accept such unorthodox solution but it just works. Furthermore it seems that fgMorphTree is designed to handle properly constant expression folding and it should work as good as Jit could for such task module CSE.

Reply:
> all of the transformations it might do (e.g. we can't split trees in the importer)

FWIW it is possible to insert new tree and to not import part of tree. I have not analyzed all importer code but this parts which are used by HW intrinsics can handle the 2 operations introduced: (i) replace not imported tree with new one, (ii) import existing unmorphed tree which exists side by side with second tree morphed by `fgMorphTree` call.

Reply:
Agree with Carol -- I think this will run into issues for more general tree shapes.

Reply:
> It is quite possible, however, from my testing it seems to work flawlessly.

> We need correct test logic which would give 100% coverage of code 

> Acceptance bar for this PR is high due to the obvious doubts of jit maintainers caused by unorthodox approach and it requires thorough proof.

I do not think the "bar" of this PR is higher than any other. Compilers and language-related software is different from most of the other guys, finite test cases are never enough for [turing complete](https://simple.wikipedia.org/wiki/Turing_complete]) language feature. The best approach is to understand its theory at first.

Reply:
> finite test cases are never enough for turing complete language feature

Pretty much false, you can have provable correct compilers and many such compilers exist. Everything revolves around automatic theorem proving and proper formalization of compiler.

Reply:
Adding lots of tests won't convince me of sufficiently low risk. There are **lots** of JIT tests, but we still run into new bugs regularly.

I am also not in favor of going with an ad-hoc approach "for now", and opening an issue later. We have many open issues, and I question whether we'd ever go back and address this.

Even approach 2, generalizing `fgMorphExpr` is not without risk, as it may call the more general morph methods. But I think that's the approach I would favor. If we can't determine that it is a reasonably "closed" method, then we can consider a more ad-hoc approach.

In any event we probably need an issue to address the need to disentangle `fgMorphTree` in such a way that 1) it has an appropriate entry point for any phase of the JIT, 2) it is less phase-dependent, and 3) any phase dependencies that are needed are explicit and maintained as first-class `Compiler` state.

Reply:
> disentangle fgMorphTree in such a way that 1) it has an appropriate entry point for any phase of the JIT, 2) it is less phase-dependent, and 3) any phase dependencies that are needed are explicit and maintained as first-class Compiler state.

`fgMorphTree` as a whole creates states which are hard to predict and analyze, however, subset of this method and calls which are made to other `fgMorphNNNN` methods can be reasonable fast analyzed and selected for use during import. My next attempt does just that by selecting group of only const nodes which fall into basic arithmetic, logical binary, logical and comparison operator categories plus integral casts. I have analyzed code paths which will be executed and have not noticed any node conversions or operations which would create unknown to importer nodes (I could have overlooked something but will do double check tomorrow).

This code does not solve the problem of `fgMorphTree` code refactoring but seems to solve the issue which this PR is trying to fix.

Reply:
Created #19932 to track refactoring of `fgMorphTree`